File: py_service_pkg/scripts/emergency_stop.py
#! /usr/bin/env python3

import rospy
from roslaunch.pmon import start_process_monitor

from geometry_msgs.msg import Twist
from std_srvs.srv import SetBool, SetBoolResponse

class EmergencyStopNode(object):

    def __init__(self):
        self.cmd_vel_pub_ = rospy.Publisher("cmd_vel", Twist, queue_size=10)
        self.stop_server_ = rospy.Service("emergency_stop", SetBool, self.stop_cb)
        
        self.pm_ = start_process_monitor()
        self.twist_msg_   = Twist()
        self.response_    = SetBoolResponse()

        rospy.loginfo("E Stop Server Started")
        
        self.twist_pub()
        rospy.sleep(0.1)
    
    def twist_pub(self):
        self.twist_msg_.linear.x  = 0.5
        self.twist_msg_.angular.z = 1.0

        self.cmd_vel_pub_.publish(self.twist_msg_)
    
    def stop_cb(self, request):
        
        if request.data is True:
            self.twist_msg_.linear.x = 0.0
            self.twist_msg_.angular.z = 0.0
            self.cmd_vel_pub_.publish(self.twist_msg_)

            self.response_.success = True
            self.response_.message = "Successfully Stopped"
        else:
            self.response_.success = False
            self.response_.message = "Stop Failed"

        return self.response_

def main():
    rospy.init_node("emergency_stop_node")
    
    e_stop_node = EmergencyStopNode()
    rospy.sleep(1.0)
    e_stop_node.twist_pub()

    rospy.spin()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: cpp_service_pkg/src/emergency_stop.cpp
#include <ros/ros.h>
#include <std_srvs/SetBool.h>
#include <geometry_msgs/Twist.h>

using SetBool = std_srvs::SetBool;

class EmergencyStopNode {
private:
    ros::ServiceServer service_;
    ros::Publisher cmd_vel_pub_;

    geometry_msgs::Twist twist_msg_;
public:
    EmergencyStopNode(ros::NodeHandle *nh){
        service_ = nh->advertiseService("emergency_stop", &EmergencyStopNode::eStopCallback, this);
        cmd_vel_pub_ = nh->advertise<geometry_msgs::Twist>("cmd_vel", 10);

        ROS_INFO_STREAM("EmergencyStopNode Started");
    }

    void moveRobot(){
        twist_msg_.linear.x = 0.5;
        twist_msg_.angular.z = 1.0;

        cmd_vel_pub_.publish(twist_msg_);
    }

    bool eStopCallback(SetBool::Request &req, SetBool::Response &res){

        if(req.data == true){
            twist_msg_.linear.x = 0.0;
            twist_msg_.angular.z = 0.0;
            cmd_vel_pub_.publish(twist_msg_);

            res.success = true;
            res.message = "Successfully Stopped";

            return true;
        } else {
            res.success = false;
            res.message = "Stop Failed";

            return false;
        }
    }
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "emergency_stop_node");
    ros::NodeHandle nh;
    EmergencyStopNode e_stop_service(&nh);

    auto start_time = ros::Time::now();
    auto cur_time   = ros::Time::now();

    while( (cur_time - start_time) < ros::Duration(3.0)){
        e_stop_service.moveRobot();
        cur_time = ros::Time::now();
    }

    ros::spin();
    ros::shutdown();

    return 0;
}